### PR TITLE
Addressing boolean RAPPOR decode failures when TRUE is ~ 0.

### PR DESCRIPTION
**Pull request by ananthr**
New function .DecodeBoolean outputs an estimate for TRUE directly from EstimateBloomCounts; bypassing maps and LASSO.

This fixes a few errors that arise when the underlying distribution for TRUE is very close
to zero and maps + LASSO fails to recover "a candidate solution."

- Updated decode.R with a new private function .DecodeBoolean to decode when inputs are boolean
- Updated decode_test.R with tests to check .DecodeBoolean
- decode_assoc.R now does not use the Other category when analyzing boolean rappor inputs (k==1)
- decode_test.R helper functions now output a little more information when test fails.

### CODE REVIEW

**Comment by ilyamironov** [comment]
Since this change is introduced to address a specific bug, it'd be a good idea to add a corresponding test to decode_test.R.

File: analysis/R/association.R [hunk 1]
```diff
@@ -353,10 +352,14 @@ ComputeDistributionEM <- function(reports, report_cohorts, maps,
     f <- var_params$f
     pstar <- (1 - f / 2) * p + (f / 2) * q
     qstar <- (1 - f / 2) * q + (f / 2) * p
+    k <- var_params$k
 
-    if (ignore_other) {
+    # Ignore other probability if either ignore_other is set or k == 1
```
Comment (andychu) [inline analysis/R/association.R]: Why not have the caller set ignore_other if k == 1, instead of putting it in this function?

File: analysis/R/decode.R [hunk 2]
```diff
@@ -276,6 +279,46 @@ Resample <- function(e) {
   list(estimates = estimates, stds = stds)
 }
 
+# Private function
+# Decode for Boolean RAPPOR inputs
+# Returns a list with attribute fit only. (Inference and other aspects
+# currently not incorporated because they're unnecessary for association.)
+.DecodeBoolean <- function(counts, params, num_reports) {
+  # Set number of cohorts to 1
```
Comment (andychu) [inline analysis/R/decode.R]: Why?  I'm not sure we need this, and the comment doesn't explain it well.

File: analysis/R/decode.R [hunk 3]
```diff
@@ -276,6 +279,46 @@ Resample <- function(e) {
...
+  params$m = 1
+  # Sums across cohorts.
+  summed_counts = colSums(counts)
```
Comment (andychu) [inline analysis/R/decode.R]: be consistent and use <- for assignment

File: analysis/R/decode.R [hunk 4]
```diff
@@ -276,6 +279,46 @@ Resample <- function(e) {
...
+
+  # With params$m = 1, EstimateBloomCounts computes Boolean RAPPOR estimates
+  # given counts summed across the cohorts.
+  es <- EstimateBloomCounts(params, summed_counts)
+
+  fit = as.data.frame(matrix(ncol = 7, nrow = 2))
+  colnames(fit) = c("string", "estimate", "std_error", "proportion",
```
Comment (andychu) [inline analysis/R/decode.R]: <-

File: analysis/R/decode.R [hunk 5]
```diff
@@ -276,6 +279,46 @@ Resample <- function(e) {
...
+               "prop_std_error", "prop_low_95", "prop_high_95")
+
+  fit$string <- c("TRUE", "FALSE")
```
Comment (andychu) [inline analysis/R/decode.R]: It would be nicer if this were c(TRUE, FALSE) -- booleans not strings.  But I suspect this is something that has to be changed all over the code.

File: analysis/R/decode.R [hunk 6]
```diff
@@ -276,6 +279,46 @@ Resample <- function(e) {
...
+  # Set number of cohorts to 1
+  params$m = 1
+  # Sums across cohorts.
+  summed_counts = colSums(counts)
+
+  # With params$m = 1, EstimateBloomCounts computes Boolean RAPPOR estimates
+  # given counts summed across the cohorts.
+  es <- EstimateBloomCounts(params, summed_counts)
+
+  fit = as.data.frame(matrix(ncol = 7, nrow = 2))
```
Comment (andychu) [inline analysis/R/decode.R]: I would initialize it without the hard-coded dimensions, e.g.

fit <- data.frame(string = c("TRUE", "FALSE"),
                           estimate = ..., )

File: analysis/R/decode.R [hunk 7]
```diff
@@ -288,6 +331,11 @@ Decode <- function(counts, map, params, alpha = 0.05,
   S <- ncol(map)  # total number of candidates
 
   N <- sum(counts[, 1])
+  if (k == 1) {
+    # Heuristic for boolean var.
+    params$m = 1
```
Comment (andychu) [inline analysis/R/decode.R]: Why duplicate this?

File: analysis/R/decode_test.R [hunk 8]
```diff
@@ -291,11 +301,52 @@ TestDecode <- function() {
                         c(report8x32, noise1), map2, distribution2, 10^6,
                         tolerance_l1 = values * 3,
                         tolerance_linf = 80)
+
+}
+
+TestDecodeBool <- function() {
+  # Testing Boolean Decode
+  values <- 2
```
Comment (andychu) [inline analysis/R/decode_test.R]: num_values?

File: analysis/R/decode_test.R [hunk 9]
```diff
@@ -291,11 +301,52 @@ TestDecode <- function() {
...
+  # 1 bit; rest of the params don't matter
+  # setting up map_bool to be consistent with the Decode API and for
+  # GenerateCounts()
+  report_bool <- list(k = 1, m = 128, h = 2)
```
Comment (andychu) [inline analysis/R/decode_test.R]: bool_params seems to make more sense

File: analysis/R/decode_test.R [hunk 10]
```diff
@@ -291,11 +301,52 @@ TestDecode <- function() {
...
+  map_bool <- matrix(c(0, 1), nrow = 128 * 1, ncol = values, byrow = TRUE)
```
Comment (andychu) [inline analysis/R/decode_test.R]: Why do we need this?  I thought it wasn't using the map anymore.

File: analysis/R/decode.R [hunk 11]
```diff
@@ -276,6 +279,46 @@ Resample <- function(e) {
...
+  # With params$m = 1, EstimateBloomCounts computes Boolean RAPPOR estimates
+  # given counts summed across the cohorts.
+  es <- EstimateBloomCounts(params, summed_counts)
+
+  fit = as.data.frame(matrix(ncol = 7, nrow = 2))
+  colnames(fit) = c("string", "estimate", "std_error", "proportion",
+               "prop_std_error", "prop_low_95", "prop_high_95")
+
+  fit$string <- c("TRUE", "FALSE")
+  fit$estimate <- c(es$estimates[[1]] * num_reports,
```
Comment (andychu) [inline analysis/R/decode.R]: I would use "ests" to be consistent with the rest of the code.  And also introduce some variables for ests$stds[[1]] and ests$estimates[[1]], which will make the code more readable.

File: analysis/R/association.R (hunk 1 above)
Comment (ananthr) [inline analysis/R/association.R]: ignore_other is set globally across both (or many) RAPPOR variables. We have to use ignore_other code for strings, but ignore it for booleans.

File: analysis/R/decode.R (hunk 7 above)
Comment (ananthr) [inline analysis/R/decode.R]: Fixed.

File: analysis/R/decode.R [hunk 12]
```diff
@@ -276,6 +279,39 @@ Resample <- function(e) {
   list(estimates = estimates, stds = stds)
 }
 
+# Private function
+# Decode for Boolean RAPPOR inputs
+# Returns a list with attribute fit only. (Inference and other aspects
+# currently not incorporated because they're unnecessary for association.)
+.DecodeBoolean <- function(counts, params, num_reports) {
+  # Boolean variables are reported without cohorts and to estimate counts,
```
Comment (ananthr) [inline analysis/R/decode.R]: I tried to explain it better, please take a look.

File: analysis/R/decode_test.R [hunk 13]
```diff
@@ -242,60 +249,106 @@ TestDecode <- function() {
   # match the ground truth. Must be close enough though.
   noise0 <- list(p = 0, q = 1, f = 0)  # no noise whatsoever
 
+  # Args are: message str, test function, # repetitions,
+  #           params, map, true pdf, # clients,
+  #           tolerances
   CheckDecodeAveAndStds("Testing Decode (1/5)", CheckDecodeHelper, 100,
-                        c(report4x2, noise0), map0, distribution0, 100,
+                        c(params_4x2, noise0), map0, distribution0, 100,
                         tolerance_l1 = 5,
```
Comment (ananthr) [inline analysis/R/decode_test.R]: I was being consistent with TestDecode. I modified both TestDecode and TestDecodeBoolean to reflect comments.

File: analysis/R/decode_test.R [hunk 14]
```diff
@@ -242,60 +249,106 @@ TestDecode <- function() {
...
                         tolerance_linf = 3)
 
   noise1 <- list(p = .4, q = .6, f = .5)  # substantial noise, very few reports
   CheckDecodeAveAndStds("Testing Decode (2/5)", CheckDecodeHelper, 100,
```
Comment (ananthr) [inline analysis/R/decode_test.R]: Fixed.

File: analysis/R/decode_test.R [hunk 15]
```diff
@@ -242,60 +249,106 @@ TestDecode <- function() {
...
-                        c(report4x2, noise1), map0, distribution0, 100,
```
Comment (ananthr) [inline analysis/R/decode_test.R]: For 2 reasons; to be consistent with Decode API and because in testing, GenerateCounts() simulates total counts using a map. I commented about this a few lines earlier but brought the comments down to this line.

File: analysis/cpp/fast_em.cc [hunk 16]
```diff
@@ -194,7 +199,10 @@ static void ExpectationMaximization(
 
     pij = new_pij;  // copy
 
-    log("max_dif: %e", em_iter, max_dif);
+    if (em_iter % 200 == 0) {
```
Comment (andychu) [inline analysis/cpp/fast_em.cc]: I changed the output of fast_em.cc in the TensorFlow branch which Tom is reviewing... so these changes will probably conflict.  It matches the R output more closely now.

I think it's better to just revert these changes since they're not really related.

File: bin/test.sh [hunk 17]
```diff
@@ -51,6 +51,9 @@ write-assoc-testdata() {
 domain,flag..HTTPS
 google.com,1
 google.com,1
+google.com,1
```
Comment (andychu) [inline bin/test.sh]: Why add these lines?  I think it makes it more confusing to interpret.  Right now it's 2/7,2/7, 3/7.  Maybe change it to 10 values, like:

30% google true
20% google false
20% yahoo true
10% yahoo false
10% bing true
10% bing false

That way it's easier to eyeball.  It's hard to do the arithmetic for /7 or /9.

File: bin/test.sh [hunk 18]
```diff
@@ -135,13 +138,18 @@ decode-assoc() {
...
     --output-dir _tmp \
     --tmp-dir _tmp \
     "$@"
 
   head _tmp/assoc-*
+
+  # Printing true results to compare against
+  echo
+  echo "==> _tmp/true_values.csv <=="
+  head _tmp/true_values.csv
```
Comment (andychu) [inline bin/test.sh]: probably "cat" instead of "head" is better since it's short

File: analysis/R/association.R [hunk 19]
```diff
@@ -377,13 +380,31 @@ ComputeDistributionEM <- function(reports, report_cohorts, maps,
...
-    cond_report_dist <- mclapply(seq(length(var_report)), function(i) {
-      cohort <- var_cohort[i]
-      #Log('Report %d, cohort %d', i, cohort)
-      bit_indices <- bit_indices_by_cohort[[cohort]]
-      GetCondProb(var_report[[i]], pstar, qstar, bit_indices,
-                  prob_other = prob_other[[cohort]])
-    }, mc.cores = num_cores)
+    # TODO(pseudorandom): check RAPPOR type more systematically instead of by
+    # checking if k == 1
+    if(k == 1) {
```
Comment (andychu) [inline analysis/R/association.R]: Can you separate this out into two functions, for boolean and string?  Because we will need another one for Basic RAPPOR, right?

File: analysis/R/decode.R [hunk 20]
```diff
@@ -276,6 +279,39 @@ Resample <- function(e) {
...
+  # first sum up counts across all cohorts and then run EstimateBloomCounts
+  # with the number of cohorts set to 1.
+  params$m <- 1  # set number of cohorts to 1
+  summed_counts <- colSums(counts)  # sum counts across cohorts
+  es <- EstimateBloomCounts(params, summed_counts)  # estimate boolean counts
+
+  est <- es$estimates[[1]]
```
Comment (andychu) [inline analysis/R/decode.R]: use "ests" to be consistent with the rest of the code

File: analysis/R/association.R (hunk 19 above)
Comment (andychu) [inline analysis/R/association.R]: style: space after "if"

**Comment by andychu** [comment]
LGTM

OK thanks for fixing this, go ahead and merge and I can update the pipeline.
